src/Microsoft.Framework.FileSystemGlobbing/Infrastructure/MatcherContext.cs

```diff
@@ -3,275 +3,180 @@
+        private IList<PatternContextBase> _excludePatternContexts;
+
         public MatcherContext(Matcher matcher, DirectoryInfoBase directoryInfo)
         {
-            Matcher = matcher;
-            DirectoryInfo = directoryInfo;
+            _root = directoryInfo;
+
+            _includePatternContexts = new List<PatternContextBase>();
             foreach (var pattern in matcher.IncludePatterns)
             {
-                if (pattern.Contains == null)
-                {
-                    IncludePatternContexts.Add(new PatternContextLinearInclude(this, pattern));
-                }
-                else
-                {
-                    IncludePatternContexts.Add(new PatternContextRaggedInclude(this, pattern));
-                }
+                _includePatternContexts.Add(pattern.Contains == null ?
+                    (PatternContextBase)new PatternContextLinearInclude(this, pattern) : new PatternContextRaggedInclude(this, pattern));
             }
+
+            _excludePatternContexts = new List<PatternContextBase>();
             foreach (var pattern in matcher.ExcludePatterns)
             {
-                if (pattern.Contains == null)
-                {
-                    ExcludePatternContexts.Add(new PatternContextLinearExclude(this, pattern));
-                }
-                else
-                {
-                    ExcludePatternContexts.Add(new PatternContextRaggedExclude(this, pattern));
-                }
+                _excludePatternContexts.Add(pattern.Contains == null ?
+                    (PatternContextBase)new PatternContextLinearExclude(this, pattern) : new PatternContextRaggedExclude(this, pattern));
             }
         }
 
-        public Matcher Matcher { get; }
-        public DirectoryInfoBase DirectoryInfo { get; }
-        public IList<PatternContextBase> IncludePatternContexts { get; } = new List<PatternContextBase>();
-        public IList<PatternContextBase> ExcludePatternContexts { get; } = new List<PatternContextBase>();
-        public List<string> Files { get; private set; }
-
-        public FrameData Frame;
-        public Stack<FrameData> FrameStack = new Stack<FrameData>();
-
         public PatternMatchingResult Execute()
         {
-            Files = new List<string>();
+            var files = new List<string>();
 
-            Frame.Stage = Stage.Complete;
-            PushFrame(Stage.Predicting, DirectoryInfo);
+            // Walk the file system in Non-recursive, pre order, tree traversal
 
-            while (Frame.Stage != Stage.Complete)
+            var stack = new Stack<FrameData>();
+            PushFrame(new FrameData(_root), stack);
+            while (stack.Count > 0)
             {
-                DoPredicting();
-                DoEnumerating();
-                DoRecursion();
+                var current = stack.Peek();
+
+                DoEnumerating(current, files);
+                DoRecursion(current, stack);
```

Comment:
It's not really quite that... DoEnumerating is when you're in the stage of deciding what are the contents of the current focus. DoRecursion is doing the actual processing of contents of that focus, and recursing is really just a part of that.